README.md
# README

# Switch Up

An animated toggle switch which can be fully customized with desired width, colors, text, corner radius, animation etc. It also maintains selection state.

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:  ...  switch_up: latest_version
```

Import it:

```dart
import 'package:switch_up/switch_up.dart';
```

## Usage Examples

```dart
// Here, default theme color are used for selected item.
SwitchUp<String>(
  backgroundColor: Colors.grey.shade200,
  items: const <String>['Home', 'Chat', 'Settings'],
  onChanged: (String value) {
    print(value);
  },
  value: 'Settings',
),
```

![ezgif.com-gif-maker(3).gif](https://github.com/ketan-glitch/switch_up/blob/708abe4f5f903dbecf82a60346d22b6e24f5894c/assets/1.gif)

### Switch Up with custom Height, Color and Animation

```dart
SwitchUp(
  backgroundColor: Colors.grey.shade200,
  color: Colors.amber,
  radius: 0,
	height: 50,
	curves: Curves.easeInOutBack,
  items: const ['Active', 'Pending'],
  onChanged: (value) {
    print(value);
  },
  value: 'Pending',
),
```

![ezgif.com-gif-maker(4).gif](https://github.com/ketan-glitch/switch_up/blob/708abe4f5f903dbecf82a60346d22b6e24f5894c/assets/2.gif)

### Switch Up with a Gradient Color

```dart
SwitchUp(
  gradient: const LinearGradient(
    begin: Alignment.topLeft,
    end: Alignment.bottomRight,
    colors: [
      Color(0xFF941F37),
      Color(0xFF590012),
    ],
  ),
  curves: Curves.bounceOut,
  backgroundColor: Colors.grey.shade200,
  items: const ['Male', 'Female', 'Other'],
  onChanged: (value) {
    print(value);
  },
  value: 'Female',
),
```

![ezgif.com-gif-maker(5).gif](https://github.com/ketan-glitch/switch_up/blob/708abe4f5f903dbecf82a60346d22b6e24f5894c/assets/3.gif)

```dart
SwitchUp(
  backgroundColor: Colors.grey.shade200,
  radius: 40,
  gradient: const RadialGradient(
    radius: 10,
    colors: [
      Color(0xFF941F37),
      Color(0xFF590012),
    ],
  ),
  curves: Curves.easeInOutQuart,
  items: const ['Normal', 'Bold', 'Italic'],
  onChanged: (value) {
    log(value);
  },
  value: 'Bold',
),
```

![ezgif.com-gif-maker(6).gif](https://github.com/ketan-glitch/switch_up/blob/708abe4f5f903dbecf82a60346d22b6e24f5894c/assets/4.gif)

Switch Up with Custom Models

```dart
// Add toString() to your Custom model class
//   @override
//   String toString() {
//     return name;
//   } 
SwitchUp<States>(
  backgroundColor: Colors.pink.withOpacity(.1),
  radius: 10,
  height: 60,`
  gradient: const LinearGradient(
    begin: Alignment.topLeft,
    end: Alignment.bottomRight,
    colors: [
      Color(0xFFFC466B),
      Color(0xFF3F5EFB),
    ],
  ),
  curves: Curves.easeInOutQuart,
  items: states,
  onChanged: (States value) {
    if (kDebugMode) {
      print(value);
    }
  },
  value: states[2],
),

```

![ezgif.com-gif-maker(7).gif](https://github.com/ketan-glitch/switch_up/blob/708abe4f5f903dbecf82a60346d22b6e24f5894c/assets/5.gif)

With icons, text and different active background colors

[Example code with explanation](https://github.com/ketan-glitch/switch_up/tree/master/example)


CHANGELOG.md
## 0.0.2

* Lint fixes

## 0.0.1

* TODO: Describe initial release.


lib/switch_up.dart
library switch_up;

import 'dart:async';
import 'dart:developer';

import 'package:flutter/material.dart';

class SwitchUp<T> extends StatefulWidget {
  const SwitchUp({
    Key? key,
    required this.items,
    required this.onChanged,
    required this.value,
    this.enabled = true,
    this.radius = 8.0,
    this.elevation = 10,
    this.curves = Curves.linear,
    this.height = 30.0,
    this.gradient,

    /// const LinearGradient(
    ///   begin: Alignment.topLeft,
    ///   end: Alignment.bottomRight,
    ///   colors: [
    ///     Color(0xFF941F37),
    ///     Color(0xFF590012),
    ///   ],
    /// )

    this.color,
    this.backgroundColor,
    this.selectedTextColor,
    this.unselectedTextColor,
    this.animationDuration = const Duration(milliseconds: 200),
  })  : assert(
          color == null || gradient == null,
          'Cannot provide both a color and a gradient\n',
        ),
        assert(
          (items.length > 1),
          'Please provide at least 2 items\n',
        ),
        super(key: key);

  /// List of Items<T>
  final List<T> items;

  /// onChanged(T value)
  ///
  /// will return selected item on tap.
  final Function(T value) onChanged;

  /// Decides if Switch is enabled or not
  final bool enabled;

  /// Animation Curve
  final Curve curves;

  /// Radius for the switch
  final double? radius;

  /// Elevation for the selected switch item
  final double elevation;

  /// Height for Switch
  final double height;

  /// Gradient for selected Switch Item.
  final Gradient? gradient;

  /// Color for Selected Switch Item
  final Color? color;

  /// backgroundColor for Switch
  final Color? backgroundColor;

  /// selectedTextColor for Switch
  final Color? selectedTextColor;

  /// unselectedTextColor for Switch
  final Color? unselectedTextColor;

  ///
  /// animationDuration for Switch
  ///
  /// default is set to `const Duration(milliseconds: 200);`
  final Duration animationDuration;

  /// Type for Switch
  final T value;

  @override
  State<SwitchUp<T>> createState() => _SwitchUpState<T>();
}

class _SwitchUpState<T> extends State<SwitchUp<T>> {
  T? _selectedItem1;
  AlignmentGeometry _alignment = const Alignment(-1, 0);

  @override
  void initState() {
    super.initState();

    ///
    /// Setting the default value
    ///
    Timer.run(() {
      setState(() {
        try {
          double x = -1;
          double res = 2 / (widget.items.length - 1);
          _selectedItem1 = widget.value;
          log('${widget.items.indexOf(_selectedItem1 as T)}');
          x += (res * widget.items.indexOf(_selectedItem1 as T));
          _alignment = Alignment(x, 0);
        } catch (e) {
          log('---- ${e.toString()} ----', name: "ERROR AT initState()");
        }
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        Size size = MediaQuery.of(context).size;
        return Container(
          width: size.width,
          height: widget.height,
          decoration: BoxDecoration(
            color: widget.backgroundColor ?? const Color(0XFFF1F1F1),
            borderRadius: BorderRadius.circular(widget.radius ?? 8),
          ),
          child: Padding(
            padding: const EdgeInsets.all(2.0),
            child: Stack(
              children: [
                AnimatedAlign(
                  alignment: _alignment,
                  curve: widget.curves,
                  duration: widget.animationDuration,
                  child: Material(
                    elevation: widget.elevation,
                    borderRadius: BorderRadius.circular(widget.radius ?? 8),
                    child: Container(
                      height: widget.height,
                      width: (constraints.maxWidth) / widget.items.length,
                      decoration: BoxDecoration(
                        borderRadius: BorderRadius.circular(widget.radius ?? 8),
                        color: (widget.color == null && widget.gradient == null) ? Theme.of(context).primaryColor : widget.color,
                        gradient: widget.gradient,
                      ),
                    ),
                  ),
                ),
                Row(
                  children: [
                    for (int i = 0; i < widget.items.length; i++)
                      Expanded(
                        child: GestureDetector(
                          onTap: () {
                            ///
                            /// Setting the tapped value
                            ///
                            if (widget.enabled) {
                              double x = -1;
                              double res = 2 / (widget.items.length - 1);
                              x += (res * i);
                              _selectedItem1 = widget.items[i];
                              try {
                                widget.onChanged((widget.items[i]));
                              } catch (e) {
                                log('---- ${e.toString()} ----', name: "ERROR AT build()");
                              }
                              _alignment = Alignment(x, 0);
                              setState(() {});
                            }
                          },
                          child: Container(
                            height: widget.height,
                            decoration: BoxDecoration(
                              borderRadius: BorderRadius.circular(8),
                            ),
                            child: Center(
                              child: Text(
                                widget.items[i].toString(),
                                style: Theme.of(context).textTheme.headline2!.copyWith(
                                      color: _selectedItem1 == widget.items[i] ? Colors.white : null,
                                      fontSize: 16.0,
                                    ),
                              ),
                            ),
                          ),
                        ),
                      ),
                  ],
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}


example/lib/main.dart
import 'dart:developer';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:switch_up/switch_up.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Switch Up',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        primaryColor: const Color(0xFFfc8e28),
      ),
      home: const MyHomePage(title: 'Switch Up'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<States> states = [
    States(id: '1', name: 'Maharashtra', country: 'India'),
    States(id: '2', name: 'Gujarat', country: 'India'),
    States(id: '3', name: 'Goa', country: 'India'),
  ];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 30.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const SizedBox(height: 50),
              SwitchUp<String>(
                backgroundColor: Colors.grey.shade200,
                items: const <String>['Home', 'Chat', 'Settings'],
                onChanged: (String value) {
                  if (kDebugMode) {
                    print(value);
                  }
                },
                value: 'Settings',
              ),
              const SizedBox(height: 50),
              SwitchUp(
                backgroundColor: Colors.grey.shade200,
                color: Colors.amber,
                radius: 0,
                height: 50,
                items: const ['Active', 'Pending'],
                onChanged: (value) {
                  log(value);
                },
                value: 'Pending',
              ),
              const SizedBox(height: 50),
              SwitchUp(
                gradient: const LinearGradient(
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  colors: [
                    Color(0xFF941F37),
                    Color(0xFF590012),
                  ],
                ),
                curves: Curves.bounceOut,
                backgroundColor: Colors.grey.shade200,
                items: const ['Male', 'Female', 'Other'],
                onChanged: (value) {
                  log(value);
                },
                value: 'Female',
              ),
              const SizedBox(height: 50),
              SwitchUp(
                backgroundColor: Colors.grey.shade200,
                radius: 40,
                gradient: const RadialGradient(
                  // begin: Alignment.topLeft,
                  // end: Alignment.bottomRight,
                  radius: 10,
                  colors: [
                    Color(0xFF941F37),
                    Color(0xFF590012),
                  ],
                ),
                curves: Curves.easeInOutQuart,
                items: const ['Normal', 'Bold', 'Italic'],
                onChanged: (value) {
                  log(value);
                },
                value: 'Bold',
              ),
              const SizedBox(height: 50),
              SwitchUp<States>(
                backgroundColor: Colors.pink.withOpacity(.1),
                radius: 10,
                height: 60,
                gradient: const LinearGradient(
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  colors: [
                    Color(0xFFFC466B),
                    Color(0xFF3F5EFB),
                  ],
                ),
                curves: Curves.easeInOutQuart,
                items: states,
                onChanged: (States value) {
                  if (kDebugMode) {
                    print(value);
                  }
                },
                value: states[2],
              ),
              const SizedBox(height: 50),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Expanded(
                    child: SwitchUp(
                      backgroundColor: Colors.grey.shade200,
                      height: 50,
                      radius: 40,
                      items: const ['Male', 'Female', 'Other'],
                      onChanged: (value) {
                        log(value);
                      },
                      value: 'Other',
                      curves: Curves.ease,
                    ),
                  ),
                  const SizedBox(width: 10),
                  Expanded(
                    child: SwitchUp(
                      backgroundColor: Colors.grey.shade200,
                      color: Colors.amber,
                      radius: 0,
                      height: 50,
                      items: const ['Active', 'Pending'],
                      onChanged: (value) {
                        log(value);
                      },
                      value: 'Pending',
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class States {
  final String id;
  final String name;
  final String country;

  States({required this.id, required this.name, required this.country});

  @override
  String toString() {
    return name;
  }
}
